README.md
<h1 align="center">
	<a href="https://sprngr.github.io/px2bfm/" target="_blank" alt="Go to Web App">
        <img src="logo.png" alt="px2bfm">
    </a>
</h1>

> Parse pixels to [BitFontMaker2](http://www.pentacom.jp/pentacom/bitfontmaker2/) JSON font format


[![Travis](https://img.shields.io/travis/sprngr/px2bfm.svg?style=for-the-badge)](https://travis-ci.org/sprngr/px2bfm)
[![npm](https://img.shields.io/npm/v/px2bfm.svg?style=for-the-badge)](https://www.npmjs.com/package/px2bfm)
![From Pittsburgh With Love](https://img.shields.io/badge/from%20pittsburgh%20with-%E2%9D%A4-yellow.svg?style=for-the-badge)

Pixels 2 [BitFontMaker2](http://www.pentacom.jp/pentacom/bitfontmaker2/) (aka px2bfm) is a simple command line tool (& [web app](https://sprngr.github.io/px2bfm/)) with the intention of
allowing you to use your desired pixel editor to draw font glyphs, creating compatible BitFontMaker2 JSON, and allowing you to import it to generate a new bitmap font. This doesn't replace the app, just supports it.

# Table of Contents

- [Web App](#web-app)
- [Installation](#installation)
	- [Alternative Installation](#alternative-installation)
- [Usage](#usage)
	- [Important Notes](#important-notes)
	- [Example Sprite Sheet & Output](#example-sprite-sheet-output)
- [Character Support](#character-support)
- [Download Template](#download-template)
- [License](#license)

# [Web App](https://sprngr.github.io/px2bfm/)

It runs exactly the same library on npm, just packaged up neatly within a GitHub hosted site. The only requirements here are to have compatible sprite sheets, template downloads
and other important usage information can be found below in this README.

# Installation
> This package requires npm and node.js. If you don't have npm or node.js on your local machine, check out [this guide on getting setup](https://docs.npmjs.com/getting-started/installing-node).

```console
$ npm install --global px2bfm
```

## Alternative Installation

If you don't feel like installing it globally via npm, you can clone this repo and run the script directly:
```console
$ ./bin/px2bfm
```

# Usage
```console
$ px2bfm --help

  Usage: px2bfm [options] <file>

  Parse pixels to BitFontMaker2 JSON font format
  For more info visit https://github.com/sprngr/px2bfm

  Options:

    -v, --version              output the version number
    -f, --fontname <fontname>  add name of the font
    -c, --creator <creator>    add name of the font creator
    -h, --help                 output usage information
```

The `fontname` and `creator` arguments are optional, but you'll probably want to set that at some point.

Example:
```console
$ px2bfm yourFileHere.png
# Output: {A lot of JSON}
```

To get this into a file:
```console
$ px2bfm yourFileHere.png > bfm2.json # Or whatever you want to call it
```

To put it on the clipboard to paste into the textfield on the web app:

**MacOS**
```console
$ px2bfm yourFileHere.png | pbcopy
```

**Windows**
```console
# I'm pretty sure this is a standard tool
$ px2bfm yourFileHere.png | clip
```

**Linux**
```console
# Let's be honest, if you're on linux you're a pro at this so why do I need to tell you
$ px2bfm yourFileHere.png | xclip
```

## Important Notes

Using one of the font templates provided (or your own, as long as it fits within required dimensions), create a pixel font.

Sprite sheets must have a width of 208px to maintain some parity with the layout of BitFontMaker2.
The height of the file must be a power of 16 (default is 128px), this will allow for future support of extended character sets.

Be sure to use black (hex #000000 / rgba 0, 0, 0) for glyph pixels, the image parser only cares about those. Any other colors & alpha values are ignored.

Each glyph must fit within a 16x16 pixel block, this is required by the web app.

## Example Sprite Sheet & Output

> Input

<img src="templates/test.png"/>

> Output

```JSON
{
	"33": [0, 0, 0, 0, 8, 8, 8, 8, 8, 8, 0, 8, 0, 0, 0, 0],
	"34": [0, 0, 0, 0, 20, 20, 20, 0, 0, 0, 0, 0, 0, 0, 0, 0],
	"35": [0, 0, 0, 0, 0, 40, 40, 124, 40, 124, 40, 40, 0, 0, 0, 0],
	...
	"name": "PixelFont",
	"copy": "PixelFontMaker",
	"letterspace": "64",
	"basefont_size": "357",
	"basefont_left": "62",
	"basefont_top": "0",
	"basefont": "None",
	"basefont2": ""
}
```

# Character Support
Currently this supports the first set of 104 characters on BitFontMaker2:
```
# Set 1
ABCDEFGHIJKLM
NOPQRSTUVWXYZ
abcdefghijklm
nopqrstuvwxyz
0123456789!"#
$%&'()*+,-./:
;<=>?@[\]^_`{
|}~¡¢£€¤¥¦§¨©
```
Extended character support is planned for the future, as well as the additional 90 character slots allowed.

# Download Template
Font template includes guides. The blue dotted line creates the grid of 16x16 blocks, the red lines are the baseline like in the web app.

<img src="templates/template.png"/>

## Downloads
* [.ase (Aseprite) format](https://github.com/sprngr/px2bfm/blob/master/templates/template.ase?raw=true)
* [.png format](https://github.com/sprngr/px2bfm/blob/master/templates/template.png?raw=true)

# License
MIT &copy; [Michael Springer](https://github.com/sprngr) 2018

BitFontMaker2 is the property of Pentacom.


CHANGELOG.md
# px2bfm Change Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/) and this project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]
### Added
- Fix for broken test
- Decoupled GitHub page to a separate branch

## [1.0.2] - 2018-03-13
### Added
- Bugfix for misplaced Euro sign (8364). It was erroneously added to the end in the mapping array.
- Static single page app implementing the package, still needs to be styled but functions.

## [1.0.1] - 2018-03-11
### Added
- Added working tests because I broke the first release

## [1.0.0] - 2018-03-11
- Initial release


test.js
import test from 'ava';

const px2bfm = require('./lib/px2bfm.js');

test('Generates BitFontMaker2 JSON', async t => {
    const fontname = 'Test Name';
    const creator = 'Test Creator';
    const fileValue = './templates/test.png';
    const outputJSON =  '{"33":[0,0,0,0,8,8,8,8,8,8,0,8,0,0,0,0],"34":[0,0,0,0,20,20,20,0,0,0,0,0,0,0,0,0],"35":[0,0,0,0,0,40,40,124,40,124,40,40,0,0,0,0],"36":[0,0,0,16,56,84,20,56,80,84,56,16,0,0,0,0],"37":[0,0,0,0,272,296,144,128,64,576,1312,544,0,0,0,0],"38":[0,0,0,0,48,72,72,48,296,196,196,312,0,0,0,0],"39":[0,0,0,0,4,4,4,0,0,0,0,0,0,0,0,0],"40":[0,0,0,0,16,8,8,4,4,4,4,8,8,16,0,0],"41":[0,0,0,0,4,8,8,16,16,16,16,8,8,4,0,0],"42":[0,0,0,0,84,56,124,56,84,0,0,0,0,0,0,0],"43":[0,0,0,0,0,0,16,16,124,16,16,0,0,0,0,0],"44":[0,0,0,0,0,0,0,0,0,0,0,8,8,4,0,0],"45":[0,0,0,0,0,0,0,0,0,28,0,0,0,0,0,0],"46":[0,0,0,0,0,0,0,0,0,0,0,8,0,0,0,0],"47":[0,0,0,0,0,16,16,16,8,8,8,4,4,4,0,0],"48":[0,0,0,0,56,100,100,84,84,76,76,56,0,0,0,0],"49":[0,0,0,0,32,48,40,32,32,32,32,32,0,0,0,0],"50":[0,0,0,0,56,68,64,32,16,8,4,124,0,0,0,0],"51":[0,0,0,0,56,68,64,48,64,64,68,56,0,0,0,0],"52":[0,0,0,0,64,96,80,72,68,252,64,64,0,0,0,0],"53":[0,0,0,0,124,4,4,60,64,64,68,56,0,0,0,0],"54":[0,0,0,0,56,68,4,60,68,68,68,56,0,0,0,0],"55":[0,0,0,0,124,64,32,32,16,16,8,8,0,0,0,0],"56":[0,0,0,0,56,68,68,56,68,68,68,56,0,0,0,0],"57":[0,0,0,0,56,68,68,68,120,64,64,56,0,0,0,0],"58":[0,0,0,0,0,0,0,8,0,0,0,8,0,0,0,0],"59":[0,0,0,0,0,0,0,8,0,0,0,8,8,4,0,0],"60":[0,0,0,0,0,0,192,48,12,48,192,0,0,0,0,0],"61":[0,0,0,0,0,0,0,124,0,124,0,0,0,0,0,0],"62":[0,0,0,0,0,0,12,48,192,48,12,0,0,0,0,0],"63":[0,0,0,0,56,68,68,32,16,16,0,16,0,0,0,0],"64":[0,0,0,0,0,480,528,1480,1320,1320,968,1040,992,0,0,0],"65":[0,0,0,0,32,32,80,80,136,248,260,260,0,0,0,0],"66":[0,0,0,0,120,136,136,120,136,136,136,120,0,0,0,0],"67":[0,0,0,0,240,264,4,4,4,4,264,240,0,0,0,0],"68":[0,0,0,0,120,136,264,264,264,264,136,120,0,0,0,0],"69":[0,0,0,0,248,8,8,120,8,8,8,248,0,0,0,0],"70":[0,0,0,0,248,8,8,120,8,8,8,8,0,0,0,0],"71":[0,0,0,0,240,264,4,4,452,260,264,240,0,0,0,0],"72":[0,0,0,0,264,264,264,504,264,264,264,264,0,0,0,0],"73":[0,0,0,0,8,8,8,8,8,8,8,8,0,0,0,0],"74":[0,0,0,0,32,32,32,32,32,32,36,24,0,0,0,0],"75":[0,0,0,0,136,72,40,24,40,72,136,264,0,0,0,0],"76":[0,0,0,8,8,8,8,8,8,8,8,248,0,0,0,0],"77":[0,0,0,520,792,792,680,680,584,584,520,520,0,0,0,0],"78":[0,0,0,0,136,152,152,168,168,200,200,136,0,0,0,0],"79":[0,0,0,0,112,136,260,260,260,260,136,112,0,0,0,0],"80":[0,0,0,0,120,136,136,136,120,8,8,8,0,0,0,0],"81":[0,0,0,0,112,136,260,260,260,260,200,496,0,0,0,0],"82":[0,0,0,0,248,264,264,264,248,72,136,264,0,0,0,0],"83":[0,0,0,0,112,136,8,48,64,128,136,112,0,0,0,0],"84":[0,0,0,0,124,16,16,16,16,16,16,16,0,0,0,0],"85":[0,0,0,0,264,264,264,264,264,264,264,240,0,0,0,0],"86":[0,0,0,0,260,260,136,136,80,80,32,32,0,0,0,0],"87":[0,0,0,0,1028,1092,1092,680,680,680,272,272,0,0,0,0],"88":[0,0,0,0,136,80,80,32,32,80,80,136,0,0,0,0],"89":[0,0,0,0,68,68,40,40,16,16,16,16,0,0,0,0],"90":[0,0,0,0,252,128,64,32,16,8,4,252,0,0,0,0],"91":[0,0,0,0,24,8,8,8,8,8,8,8,8,24,0,0],"92":[0,0,0,0,4,4,4,8,8,8,16,16,16,0,0,0],"93":[0,0,0,0,12,8,8,8,8,8,8,8,8,12,0,0],"94":[0,0,0,0,16,40,40,68,68,0,0,0,0,0,0,0],"95":[0,0,0,0,0,0,0,0,0,0,0,0,0,252,0,0],"96":[0,0,0,0,4,8,0,0,0,0,0,0,0,0,0,0],"97":[0,0,0,0,0,0,0,56,68,120,68,120,0,0,0,0],"98":[0,0,0,0,8,8,8,8,120,136,136,120,0,0,0,0],"99":[0,0,0,0,0,0,0,56,68,4,68,56,0,0,0,0],"100":[0,0,0,0,64,64,64,120,68,68,68,120,0,0,0,0],"101":[0,0,0,0,0,0,0,56,68,124,4,120,0,0,0,0],"102":[0,0,0,0,16,8,8,28,8,8,8,8,0,0,0,0],"103":[0,0,0,0,0,0,0,56,68,68,68,120,64,56,0,0],"104":[0,0,0,0,8,8,8,120,136,136,136,136,0,0,0,0],"105":[0,0,0,0,0,8,0,8,8,8,8,8,0,0,0,0],"106":[0,0,0,0,0,8,0,8,8,8,8,8,8,4,0,0],"107":[0,0,0,0,8,8,8,72,40,24,40,72,0,0,0,0],"108":[0,0,0,0,8,8,8,8,8,8,8,8,0,0,0,0],"109":[0,0,0,0,0,0,0,440,584,584,584,584,0,0,0,0],"110":[0,0,0,0,0,0,0,120,136,136,136,136,0,0,0,0],"111":[0,0,0,0,0,0,0,56,68,68,68,56,0,0,0,0],"112":[0,0,0,0,0,0,0,120,136,136,136,120,8,8,0,0],"113":[0,0,0,0,0,0,0,120,68,68,68,120,64,64,0,0],"114":[0,0,0,0,0,0,0,104,24,8,8,8,0,0,0,0],"115":[0,0,0,0,0,0,0,56,4,24,32,28,0,0,0,0],"116":[0,0,0,0,0,8,8,28,8,8,8,16,0,0,0,0],"117":[0,0,0,0,0,0,0,136,136,136,136,240,0,0,0,0],"118":[0,0,0,0,0,0,0,68,68,40,40,16,0,0,0,0],"119":[0,0,0,0,0,0,0,292,292,340,136,136,0,0,0,0],"120":[0,0,0,0,0,0,0,68,40,16,40,68,0,0,0,0],"121":[0,0,0,0,0,0,0,68,68,40,40,16,16,12,0,0],"122":[0,0,0,0,0,0,0,124,32,16,8,124,0,0,0,0],"123":[0,0,0,0,16,8,8,8,8,4,8,8,8,8,16,0],"124":[0,0,0,0,8,8,8,8,8,8,8,8,8,8,0,0],"125":[0,0,0,0,4,8,8,8,8,16,8,8,8,8,4,0],"126":[0,0,0,0,0,0,0,80,40,0,0,0,0,0,0,0],"161":[0,0,0,0,0,0,8,0,8,8,8,8,8,8,0,0],"162":[0,0,0,0,0,0,64,56,100,20,76,56,4,0,0,0],"163":[0,0,0,112,136,136,8,60,8,8,4,252,0,0,0,0],"164":[0,0,0,0,68,56,40,56,68,0,0,0,0,0,0,0],"165":[0,0,0,0,68,68,40,40,16,56,56,16,0,0,0,0],"166":[0,0,0,0,8,8,8,8,0,0,8,8,8,8,0,0],"167":[0,0,0,0,0,112,136,8,112,136,112,128,136,112,0,0],"168":[0,0,0,0,20,0,0,0,0,0,0,0,0,0,0,0],"169":[0,0,0,0,240,264,612,532,532,612,264,240,0,0,0,0],"8364":[0,0,0,0,224,16,8,124,124,8,16,224,0,0,0,0],"name":"Test Name","copy":"Test Creator","letterspace":"64","basefont_size":"357","basefont_left":"62","basefont_top":"0","basefont":"None","basefont2":""}';

    await Promise.resolve(px2bfm(fileValue, fontname, creator))
        .then((output) => {
            t.is(outputJSON, output);
        });
});


lib/px2bfm.js
'use strict';

// Required Modules
// Browser support requires usage of jimp/browser
const Jimp = (typeof window != 'undefined') ? window.Jimp : require('jimp');

// bitmask values for each column
const bitMaskMap = [1,2,4,8,16,32,64,128,256,512,1024,2048,4096,8192,16384,32768];

// Currently supports first set of glyphs BitFontMaker2, extended set soon
const glyphMapLayout = require('./glyphMap.json');

// Size of glyph canvas, 16x16 is used by BitFontMaker2
const glyphSize = 16;

// Boilerplate output JSON
const bfmJSON = require('./bitfontmaker2.json');
let outputJSON;

// Default dimensions; width is constant, height can be changed by user input
let glyphMapDimensions = {
    height: 128,
    width: 208
}

// Output JSON of glyphs for BitFontMaker2
let glyphMapOutput = {};

// Entry point of the library
//
// imagePath - the path of the image to be processed
// name - name of the font, defaults to "PixelFont"
// creator - name of the font creator, defaults to "PixelFontMaker"
//
// Returns promise which resolves as the font JSON string
module.exports = function (imagePath, name, creator) {
    // Reset outputJSON on every run with deep copy
    // This is specifically for the browser
    outputJSON = JSON.parse(JSON.stringify(bfmJSON));

    // Set font information
    outputJSON.name = name || "PixelFont";
    outputJSON.copy = creator || "PixelFontMaker";

    return Jimp.read(imagePath)
        .then((image) => {
            let height = image.bitmap.height;
            let width = image.bitmap.width;

            // Check width
            if (width != glyphMapDimensions.width) {
                console.error('Image must be 208 pixels wide (13 16x16 glyphs across)');
                return;
            }

            // Check height (could be more or less)
            if (height % glyphSize !== 0) {
                console.error('Glyph heights must be a power of 16 (128, 256, etc)');
                return;
            } else if (height !== glyphMapDimensions.height) {
                glyphMapDimensions.height = height;
            }

            return processPixels(getPixelsFromBuffer(image.bitmap.data));
        })
        .catch(function (err) {
            console.error(err);
            return;
        });
}

// Converts Buffer into array of pixels
//
// buffer - Buffer created from jimp of input image
//
// Returns array of objects containing rgba values
function getPixelsFromBuffer(buffer) {
    let pixelData = [...buffer];
    let results = [];

    for (let i = 0; i < pixelData.length; i += 4) {
      results.push({
          r: pixelData[i+0],
          g: pixelData[i+1],
          b: pixelData[i+2],
          a: pixelData[i+3]
      })
    }
    return results;
}

// Processes value stored at glyphPixels
function processPixels(pixels) {
    let outerIndex = {x:-1, y: 0};
    let innerIndex = {x:0, y: 0};

    pixels.some((pixel, index) => {
        // Set inner x to be 0 - 15 based on modulo of current index
        innerIndex.x = index % glyphSize;

        // if inner x is 0, increase outer x
        if (innerIndex.x == 0) {
            outerIndex.x++;

            // if outer x will be out of bounds (13)
            // reset it to 0, bump the inner y
            if (outerIndex.x == glyphMapDimensions.width / glyphSize) {
                outerIndex.x = 0;
                innerIndex.y++;

                // if inner y is out of bounds (16)
                // reset it to 0, bump the outer y
                if (innerIndex.y == glyphSize) {
                    innerIndex.y = 0;
                    outerIndex.y++;
                }
            }
        }

        // Math it up. If rgb is 0, it's part of the glyph.
        if (pixel.r + pixel.g + pixel.b == 0) {
            outputJSON[glyphMapLayout[outerIndex.y][outerIndex.x]][innerIndex.y] += bitMaskMap[innerIndex.x];
        }
    });

    return JSON.stringify(outputJSON);
}


lib/px2bitfontmaker2.js
'use strict';
const Jimp = require('jimp');

const bitMaskMap = [1,2,4,8,16,32,64,128,256,512,1024,2048,4096,8192,16384,32768];

// Currently supports first set of glyphs bitfontmaker2, extended set soon
const glyphMapLayout = require('./glyphMap.json');
const glyphSize = 16;

let outputJSON = require('./bitfontmaker2.json');

// Import image, get pixels
let glyphPixels;
let glyphMapDimensions = {
    height: 128,
    width: 208
}
let glyphMapOutput = {};

module.exports = function px2bfm(imagePath, name, creator) {
    // Set font information
    outputJSON.name = name || "PixelFont";
    outputJSON.copy = creator || "PixelFontMaker";

    return Jimp.read(imagePath)
        .then((image) => {
            let height = image.bitmap.height;
            let width = image.bitmap.width;

            // Check width
            if (width != glyphMapDimensions.width) {
                console.error('Image must be 208 pixels wide (13 16x16 glyphs across)');
                return;
            }

            // Check height (could be more or less)
            if (height % glyphSize !== 0) {
                console.error('Glyph heights must be a power of 16 (128, 256, etc)');
                return;
            } else if (height !== glyphMapDimensions.height) {
                glyphMapDimensions.height = height;
            }

            glyphPixels = getPixelsFromBuffer(image.bitmap.data);

            return processPixels();
        })
        .catch(function (err) {
            console.error(err)
            return;
        });
}

function getPixelsFromBuffer(buffer) {
    let pixelData = [...buffer];
    let results = [];

    for (let i=0; i<pixelData.length; i+=4) {
      results.push({
          r: pixelData[i+0],
          g: pixelData[i+1],
          b: pixelData[i+2],
          a: pixelData[i+3]
      })
    }
    return results;
}

function processPixels() {
    let outerIndex = {x:-1, y: 0};
    let innerIndex = {x:0, y: 0};

    glyphPixels.some((pixel, index) => {
        // Set inner x to be 0 - 15 based on modulo of current index
        innerIndex.x = index % glyphSize;

        // if inner x is 0, increase outer x
        if (innerIndex.x == 0) {
            outerIndex.x++;

            // if outer x will be out of bounds (13)
            // reset it to 0, bump the inner y
            if (outerIndex.x == glyphMapDimensions.width / glyphSize) {
                outerIndex.x = 0;
                innerIndex.y++;

                // if inner y is out of bounds (16)
                // reset it to 0, bump the outer y
                if (innerIndex.y == glyphSize) {
                    innerIndex.y = 0;
                    outerIndex.y++;
                }
            }
        }

        if (pixel.r == 0 && pixel.g == 0 && pixel.b == 0) {
            outputJSON[glyphMapLayout[outerIndex.y][outerIndex.x]][innerIndex.y] += bitMaskMap[innerIndex.x];
        }
    });

    return JSON.stringify(outputJSON);
}


assets/javascript/app.js
'use strict';

window.Buffer = require('buffer');
const px2bfm = require('px2bfm');

// Drag and Drop code based off Smashing Magazine
// https://www.smashingmagazine.com/2018/01/drag-drop-file-uploader-vanilla-js/
let dropArea = document.querySelector('#drop-area');
let outputTextarea = document.querySelector('#output');
let outputRaw = document.querySelector('#output-raw');
let fontNameInput = document.querySelector('#font-name');
let fontCreatorInput = document.querySelector('#font-creator');
let log = document.querySelector('#error-display');

// Override console.error to let it bubble up properly
['error'].forEach((verb) => {
    console[verb] = ((method, verb, log) => {
        return (text) => {
            method(text);
            var msg = document.createElement('code');
            msg.textContent = 'Error: ' + text;
            log.appendChild(msg);
            log.classList.remove('hidden');
        };
    })(console[verb].bind(console), verb, log);
});

// Prevent default drag behaviors
['dragenter', 'dragover', 'dragleave', 'drop'].forEach(eventName => {
    dropArea.addEventListener(eventName, preventDefaults, false);
});

// Highlight drop area when item is dragged over it
['dragenter', 'dragover'].forEach(eventName => {
    dropArea.addEventListener(eventName, highlight, false);
});

['dragleave', 'drop'].forEach(eventName => {
    dropArea.addEventListener(eventName, unhighlight, false);
});

// Handle dropped files
dropArea.addEventListener('drop', handleDrop, false);

// Copy JSON on click
outputTextarea.addEventListener('click', copyText, false);

function copyText() {
    if (outputRaw.innerHTML === "") return;

    outputRaw.select();
    document.execCommand("copy");
    document.querySelector('.output-container').classList.add('active');
}

function preventDefaults (e) {
    e.preventDefault();
    e.stopPropagation();
}

function highlight(e) {
    dropArea.classList.add('active');
}

function unhighlight(e) {
    dropArea.classList.remove('active');
}

function handleDrop(e) {
    var dt = e.dataTransfer;
    var files = dt.files;

    handleFiles(files);
}

function handleFiles(files) {
    files = [...files];
    resetLogDisplay();
    files.forEach(convertFile);
    files.forEach(previewFile);
}

function convertFile(file) {
    let reader = new FileReader();
    reader.readAsDataURL(file);
    reader.onloadend = function() {
        Promise.resolve(px2bfm(reader.result, fontNameInput.value, fontCreatorInput.value))
        .then((output) => {
            // If output is undefined, an error was thrown. Handled by console override
            if (typeof output === 'undefined') return false;

            dropArea.classList.add('active');
            document.querySelector('#output-display').classList.remove('hidden');
            outputRaw.innerHTML = output;
            outputTextarea.innerHTML = output;
            hljs.highlightBlock(outputTextarea);
        });
    }
}

function previewFile(file) {
    let reader = new FileReader();
    reader.readAsDataURL(file);
    reader.onloadend = function() {
        let upload = document.querySelector('.upload img');
        upload.src = reader.result;
    }
}

function resetLogDisplay() {
    log.classList.add('hidden');
    while (log.firstChild) {
        log.removeChild(log.firstChild);
    }
}

// Exposing on window due to some browserify scoping fuckery
window.copyText = copyText;
window.handleFiles = handleFiles;


index.js
const Jimp = require('jimp');

const bitMaskMap = [1,2,4,8,16,32,64,128,256,512,1024,2048,4096,8192,16384,32768];

// Currently supports first set of glyphs bitfontmaker2, extended set soon
const glyphMapLayout = [
    [65,66,67,68,69,70,71,72,73,74,75,76,77],
    [78,79,80,81,82,83,84,85,86,87,88,89,90],
    [97,98,99,100,101,102,103,104,105,106,107,108,109],
    [110,111,112,113,114,115,116,117,118,119,120,121,122],
    [48,49,50,51,52,53,54,55,56,57,33,34,35],
    [36,37,38,39,40,41,42,43,44,45,46,47,58],
    [59,60,61,62,63,64,91,92,93,94,95,96,123],
    [124,125,126,161,162,163,164,165,166,167,168,169,8364],
];
const glyphSize = 16;

let outputJSON = require('./templates/bitfontmaker2.json');

// Import image, get pixels
let glyphPixels;
let glyphMapDimensions = {
    height: 128,
    width: 208
}
let glyphMapOutput = {};

Jimp.read('./templates/test.png')
    .then((image) => {
        let height = image.bitmap.height;
        let width = image.bitmap.width;

        // Check width
        if (width != glyphMapDimensions.width) {
            console.error('Image must be 208 pixels wide (13 16x16 glyphs across)');
            return;
        }

        // Check height (could be more or less)
        if (height % glyphSize !== 0) {
            console.error('Glyph heights must be a power of 16 (128, 256, etc)');
            return;
        } else if (height !== glyphMapDimensions.height) {
            glyphMapDimensions.height = height;
        }

        glyphPixels = getPixelsFromBuffer(image.bitmap.data);

        processPixels();
    })
    .catch(function (err) {
        console.error(err)
        return;
    });

function getPixelsFromBuffer(buffer) {
    let pixelData = [...buffer];
    let results = [];

    for (let i=0; i<pixelData.length; i+=4) {
      results.push({
          r: pixelData[i+0],
          g: pixelData[i+1],
          b: pixelData[i+2],
          a: pixelData[i+3]
      })
    }
    return results;
}

function processPixels() {
    let outerIndex = {x:-1, y: 0};
    let innerIndex = {x:0, y: 0};

    glyphPixels.some((pixel, index) => {
        // Set inner x to be 0 - 15 based on modulo of current index
        innerIndex.x = index % glyphSize;

        // if inner x is 0, increase outer x
        if (innerIndex.x == 0) {
            outerIndex.x++;

            // if outer x will be out of bounds (13)
            // reset it to 0, bump the inner y
            if (outerIndex.x == glyphMapDimensions.width / glyphSize) {
                outerIndex.x = 0;
                innerIndex.y++;

                // if inner y is out of bounds (16)
                // reset it to 0, bump the outer y
                if (innerIndex.y == glyphSize) {
                    innerIndex.y = 0;
                    outerIndex.y++;
                }
            }
        }

        if (pixel.r == 0 && pixel.g == 0 && pixel.b == 0) {
            outputJSON[glyphMapLayout[outerIndex.y][outerIndex.x]][innerIndex.y] += bitMaskMap[innerIndex.x];
        }
    });

    console.log(JSON.stringify(outputJSON));
}
